### ISSUE TITLE
Platform issues with aria-live

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/forces-and-motion-basics#145]
This issue has been observed in all sims that utilize the aria-live attribute.  It is open in this repository for documentation as it is simulation specific.

The ARIA attribute ```aria-live``` is very useful for notifying AT that an element has changed even when it is outside of focus.  However, it seems to behave unpredictably across various platforms.  For instance, with NVDA and Chrome, aria-live events seem to be triggered multiple times as an updated paragraph element can be read 2-5 times.  Other times, the paragraph can be read multiple times and then cut off half way through the reading.  In VoiceOver, the ordering of aria events seems to be reversed from other platforms.

At this time, it is unclear whether this issue is related to our instrumentation of accessible content or due to browsers in combination with AT that are not handling the aria events correctly.

Examples of aria-live in code can be found in CartNode.js and NetForceScreenView.js.  The associated code blocks are pasted here for convenience:
In CartNode.js:
```js
        /* parallel DOM element should look like:
         *
         * <div id="cart" tabIndex="-1" aria-live="assertive" >
         *   <p>The cart is stationary</p>
         *</div>
         */
        var domElement = document.createElement( 'img' );
        domElement.tabIndex = '-1';
        domElement.setAttribute( 'aria-live', 'assertive' );

        var descriptionElement = document.createElement( 'p' );
        descriptionElement.innerText = 'The cart is stationary';

        domElement.appendChild( descriptionElement );

        thisNode.cart.xProperty.link( function( x ) {
          // when the cart has moved by the value below, update the description and notify the user with aria-live
          if ( Math.abs( thisNode.xPosition - x ) > 15 ) {
            var directionString = ( thisNode.xPosition - x ) > 0 ? leftDescriptionString : rightDescriptionstring;
            thisNode.updateLiveCartRegion( descriptionElement, directionString );
            thisNode.xPosition = x;
          }
        } );
```
```js
    updateLiveCartRegion: function( descriptionElement, directionString ) {
      // update the aria live region by setting its inner text.
      descriptionElement.innerText = 'Cart accelerating ' + directionString;
    }
```
In NetForceScreenView.js:
```js
        var actionElement = document.createElement( 'p' );
        actionElement.innerText = '';
        actionElement.setAttribute( 'aria-live', 'polite' );
        actionElement.id = 'netForceActionElement';
        accessiblePeer.domElement.appendChild( actionElement );
```
Update the live actionElement in PullerNode.js
```js
                // update the live description for the net force screen
                var actionElement = document.getElementById( 'netForceActionElement' );
                var actionString = selectedDescriptionString + accessibleDescription;
                actionElement.innerText = actionString;
```

**Comment by jessegreenberg** [comment]
For reference, here are two interesting pieces of documentation:

From Chrome dev docs:
https://developer.chrome.com/extensions/a11y

Most notably:
> Because ARIA support is still a work in progress, Google Chrome might not raise an event for every ARIA property, and screen readers might not recognize all of the events being raised.

From Web Accessibility:
https://www.webaccessibility.com/best_practices.php?technology_platform_id=121

The above doc seems like a good starting point to learn about good aria-live practices.

**Comment by jessegreenberg** [comment]
In phetsims/scenery#450, I am revisiting the accessibility DOM prototype to investigate possible issues with aria-live.  In that new prototype, I am using aria-live in the following way:
Element with aria-live attribute:
```html
    <!--Element used to alert the user that an action or event has occured in the simulation-->
    <p><span id="ariaActionElement" role="alert" aria-atomic="true"></span></p>
```
script that updates the element:
```js
function placePullerOnKnot( event, puller, knot ) {
  if ( event.keyCode === 13 ) {
    var textNode = document.createTextNode( puller.alt + 'selected for drag and drop.' );
    var targetNode = document.getElementById( 'ariaActionElement' );
    while ( targetNode.firstChild ) {
      targetNode.removeChild( targetNode.firstChild );
    }
    targetNode.appendChild( textNode );
  }
}
```
With this implementation, I am not able to reproduce the issues of string repetition.  However, I notice that the aria-event will fire **only** when the text changes in some way.  So aria-live cannot be used to to let the user know that an element is *not* changing.  This is unfortunate.

**Comment by jessegreenberg** [comment]
Important to note that testing done thus far has been on Windows 7 + Chrome + NVDA.  Will need to test further, but on this platform, I am not observing any repetition of event reading.

**Comment by samreid** [comment]
> So aria-live cannot be used to to let the user know that an element is not changing.

Perhaps we could use a strategy like changing "The net force is 15 Newtons" to "The net force is still 15 Newtons" to inform the user that it explicitly hadn't changed?

**Comment by jessegreenberg** [comment]
Yes, exactly, the solution will need to be in the design of the accessible auditory description to let the user know that some property hasn't changed.

**Comment by jessegreenberg** [comment]
A more current investigation of this issue is being tracked here:
https://github.com/phetsims/balloons-and-static-electricity/issues/106

Closing this issue.
